Added SNS Acceleration IK Solver Interface with UnitTest

robodreamer opened the pull request:
- added sns acceleration ik class and SNS_Base_Acc type
- added Jacobian dot computation
- created unit test with sawyer model
- fixed some bugs in sns_ik_acc_base and sns_ik_vel_base codes

NOTE: Acc IK solver was created as another type of Vel IK solver,
I am trying to figure out a better way to do this, so any suggestions
are more than welcome! But this code works and it passes unit test.

Commit:
Added SNS Acceleration IK Solver Interface with UnitTest

- added sns acceleration ik class and SNS_Base_Acc type
- added Jacobian dot computation
- created unit test with sawyer model

Commit:
Added a missing unit test file that caused Travis to fail

Commit:
Correct a log level in the code

chris-smith commented on sns_ik_lib/src/sns_acceleration_ik.cpp:
```diff
@@ -0,0 +1,190 @@
+
+  if (n_tasks != ntasks) {
+    n_tasks = ntasks;
+    Eigen::VectorXd ddq = Eigen::VectorXd::Zero(n_dof);
```
unnecessary? Or is this supposed to set a member variable?

robodreamer replied:
It is a member variable. I followed the original code (SNS velocity solvers)'s style, so it is hard to tell whether it is a local variable or not. We should revisit all these code once and revise the style..

chris-smith commented on sns_ik_lib/src/sns_acceleration_ik.cpp:
```diff
@@ -0,0 +1,190 @@
+    // for the minimum bound
+    max = -maxJointAcceleration(i);
+    if (m_usePositionLimits) {
+      step = 2 * (jointLimit_low(i) - actualJointConfiguration(i) - actualJointVelocities(i)*loop_period) / (loop_period * loop_period);
```
is this taken from something specific I'm not aware of? worth a comment?

robodreamer replied:
It is based on the equations in the original paper. I will add a comment.

chris-smith commented on sns_ik_lib/src/sns_ik.cpp:
```diff
@@ -362,6 +373,230 @@ bool SNS_IK::nullspaceBiasTask(const KDL::JntArray& q_bias,
+                         const KDL::JntArray& q_bias,
+                         const std::vector<std::string>& biasNames,
+                         const KDL::JntArray& q_acc_bias,
+                         KDL::JntArray& qddot_out)
```
should output variable be a pointer? That looks like the convention

robodreamer replied:
Yes. I followed the original code style, which is not necessarily good but it seemed to require  a large amount of additional code changes once changed.

chris-smith reviewed (commented):
Not sure I know the code well enough to provide useful suggestions RE your note...

forrest-rm commented on sns_ik_lib/include/sns_ik/sns_acceleration_ik.hpp:
```diff
@@ -0,0 +1,104 @@
+    // The control loop period in seconds
+    void setLoopPeriod(double period) { loop_period = period; }
+
+    // SNS Acceleration IK
```
This function needs more documentation for inputs and outputs

forrest-rm replied:
Overall, more documentation is still needed.

forrest-rm commented on sns_ik_lib/include/sns_ik/sns_acceleration_ik.hpp:
```diff
@@ -0,0 +1,104 @@
+
+    // SNS Acceleration IK
+    int getJointAcceleration(Eigen::VectorXd *jointAcceleration, const std::vector<TaskAcc> &sot,
+                                    const Eigen::VectorXd &jointConfiguration, const Eigen::VectorXd &jointVelocities);
```
Check alignment with previous line

forrest-rm commented on sns_ik_lib/include/sns_ik/sns_acceleration_ik.hpp:
```diff
@@ -0,0 +1,104 @@
+    int getJointAcceleration(Eigen::VectorXd *jointAcceleration, const std::vector<TaskAcc> &sot,
+                                    const Eigen::VectorXd &jointConfiguration, const Eigen::VectorXd &jointVelocities);
+
+    std::vector<double> getTasksScaleFactor()
```
The getters should be const functions: 
getTaskScaleFactor() const {

robodreamer replied:
Good point. Fixed it.

forrest-rm commented on sns_ik_lib/include/sns_ik/sns_acceleration_ik.hpp:
```diff
@@ -0,0 +1,104 @@
+    Eigen::ArrayXd ddotQmax;  // higher joint velocity bound
+    std::vector<double> scaleFactors;
+
+    // variables for base acc ik solver
```
It looks like most of these variables are only used in one function. They should not be member variables.

robodreamer replied:
That is right. I moved them inside the function now.

forrest-rm replied:
👍

forrest-rm commented on sns_ik_lib/include/sns_ik/sns_acceleration_ik.hpp:
```diff
@@ -0,0 +1,104 @@
+
+    SnsAccIkBase::uPtr baseIkSolver;
+
+    SnsIkBase::ExitCode exitCode;
```
Same here

forrest-rm commented on sns_ik_lib/include/sns_ik/sns_ik.hpp:
```diff
@@ -146,6 +150,46 @@ namespace sns_ik {
+
+    bool getJacobianDot(const KDL::JntArray& jnt_pos_in, const KDL::JntArray& jnt_vel_in,
+                        Eigen::MatrixXd *jacobianDotOut)
+    { return getJacobianDot(jnt_pos_in, jnt_vel_in, jacobianDotOut, "right_hand"); }
```
Instead of making a new function definition, just set the default parameter for tipName above.

robodreamer replied:
That's right.

forrest-rm commented on sns_ik_lib/src/sns_acc_ik_base.cpp:
```diff
@@ -172,10 +168,7 @@ SnsIkBase::ExitCode SnsAccIkBase::solve(const Eigen::MatrixXd& J, const Eigen::V
       ddqNull(jntIdx) = (getLowerBounds())(jntIdx);
-    } else {
-      ROS_ERROR("Internal error in computing task scale!  ddq(%d) = %f", jntIdx, (*ddq)(jntIdx));
-      return ExitCode::InternalError;
```
Why was this check removed?

robodreamer replied:
From unit testing, I found out that this case was unnecessary and it caused a few tasks to fail by terminating the algorithm earlier.

forrest-rm commented on sns_ik_lib/src/sns_acceleration_ik.cpp:
```diff
@@ -0,0 +1,190 @@
+    // for the minimum bound
+    max = -maxJointAcceleration(i);
+    if (m_usePositionLimits) {
+      step = 2 * (jointLimit_low(i) - actualJointConfiguration(i) - actualJointVelocities(i)*loop_period) / (loop_period * loop_period);
```
Check line length

forrest-rm commented on sns_ik_lib/src/sns_acceleration_ik.cpp:
```diff
@@ -0,0 +1,190 @@
+    // for the maximum bound
+    max = maxJointAcceleration(i);
+    if (m_usePositionLimits) {
+      step = 2 * (jointLimit_high(i) - actualJointConfiguration(i) - actualJointVelocities(i)*loop_period) / (loop_period * loop_period);
```
Check line length

forrest-rm commented on sns_ik_lib/src/sns_acceleration_ik.cpp:
```diff
@@ -0,0 +1,190 @@
+int SNSAccelerationIK::getJointAcceleration(Eigen::VectorXd *jointAcceleration,
+    const std::vector<TaskAcc> &sot, const Eigen::VectorXd &jointConfiguration, const Eigen::VectorXd &jointVelocities)
+{
+  // This will only reset member variables if different from previous values
```
Check that sot.size() > 0

robodreamer replied:
Good suggestion!

forrest-rm commented on sns_ik_lib/src/sns_acceleration_ik.cpp:
```diff
@@ -0,0 +1,190 @@
+
+  if (exitCode != SnsIkBase::ExitCode::Success)
+  {
+    return -1;
```
Is there a reason not to return the actual exitCode?

robodreamer replied:
It was because all other existing codes were written based on integer return values.

forrest-rm commented on sns_ik_lib/src/sns_ik.cpp:
```diff
@@ -231,13 +234,21 @@ bool SNS_IK::setVelocitySolveType(VelocitySolveType type) {
         ROS_INFO("SNS_IK: Set Velocity solver to Base SNS solver.");
         break;
+      case sns_ik::SNS_Base_Acc:
+        m_ik_vel_solver = std::shared_ptr<SNSVelIKBaseInterface>(new SNSVelIKBaseInterface(m_chain.getNrOfJoints(), m_loopPeriod));
```
The end user should be able to set the velocity solver type independent of the acceleration solver type. I recommend writing a new function for setting the acceleration solver type. 

Or since, there's only one acceleration solver for now, just set it by default.

forrest-rm commented on sns_ik_lib/src/sns_ik.cpp:
```diff
@@ -362,6 +373,230 @@ bool SNS_IK::nullspaceBiasTask(const KDL::JntArray& q_bias,
+  std::vector<KDL::Frame> transforms;
+  KDL::ChainFkSolverPos_recursive fwdKin(m_chain);
+  KDL::JntArray q(jnt_pos_in.rows());
+  q.data << 1,1,1,1,1,1,1;
```
Don't hardcode the number of joints

robodreamer replied:
This was only set here for testing. I should have removed it earlier. Thanks for pointing it out.

forrest-rm commented on sns_ik_lib/src/sns_ik.cpp:
```diff
@@ -362,6 +373,230 @@ bool SNS_IK::nullspaceBiasTask(const KDL::JntArray& q_bias,
+  KDL::JntArray q(jnt_pos_in.rows());
+  q.data << 1,1,1,1,1,1,1;
+
+  for (size_t i = 0; i < 9; i++) {
```
Where does 9 come from?

robodreamer replied:
9 came from the number of links in the chain. When I implemented this in kinematicsTools code, it was easily available through helper functions, but it was tricky in sns_ik code. I will try again to find a way to get it from urdf chain.

forrest-rm commented on sns_ik_lib/src/sns_ik.cpp:
```diff
@@ -362,6 +373,230 @@ bool SNS_IK::nullspaceBiasTask(const KDL::JntArray& q_bias,
+  }
+
+  // get the tip position vector
+  Eigen::Vector3d p_0_k(transforms[8].p(0),transforms[8].p(1),transforms[8].p(2));
```
Where does 8 come from?

robodreamer replied:
I shouldn't have hard coded this, but it corresponds to the right_hand tip frame.

forrest-rm commented on sns_ik_lib/src/sns_ik.cpp:
```diff
@@ -362,6 +373,230 @@ bool SNS_IK::nullspaceBiasTask(const KDL::JntArray& q_bias,
+    w.col(i) = w_i;
+
+    // compute Jdw
+    Eigen::Vector3d Jw_i = Jw.col(i);
```
There is a lot of extra copying of vectors going on. It adds extra computations and makes this function a bit hard to read.

Just set "Jdw.col(i) = w.col(i-1).cross(Jw.col(i))". There are a few other similar temp variables. I also recommend handling i==0 outside of the loop to remove that check within the loop.

robodreamer replied:
You are right. Good suggestion. I initially tried to do it without extra variables but there were some errors so I wrote this way to avoid those. I will try it again and see if I can shorten them.

forrest-rm commented on sns_ik_lib/src/sns_ik.cpp:
```diff
@@ -362,6 +373,230 @@ bool SNS_IK::nullspaceBiasTask(const KDL::JntArray& q_bias,
+    }
+
+    // compute its derivative
+    Eigen::Vector3d p_0_i_1_i_ = p_0_i_1_i.col(i);
```
Another temp variable.

forrest-rm commented on sns_ik_lib/src/sns_ik.cpp:
```diff
@@ -362,6 +373,230 @@ bool SNS_IK::nullspaceBiasTask(const KDL::JntArray& q_bias,
+      pdot_0_i_1_k.col(i) = pdot_0_i_1_k.col(i) + pdot_0_i_1_i.col(j);
+    }
+    // compute Jdv
+    Eigen::Vector3d Jw_i = Jw.col(i);
```
Same here

forrest-rm commented on sns_ik_lib/src/sns_ik.cpp:
```diff
@@ -362,6 +373,230 @@ bool SNS_IK::nullspaceBiasTask(const KDL::JntArray& q_bias,
+  task.jacobian = jacobian.data;
+  Eigen::VectorXd dJdq = jacobianDot*qdot_in.data;
+  task.dJdq = dJdq;
+  task.desired = Eigen::VectorXd::Zero(6);
```
6 is another magic number. I believe number of Cartesian dimensions is defined somewhere already.

robodreamer replied:
I couldn't find it in sns_ik code. I defined a new one now.

forrest-rm commented on sns_ik_lib/src/sns_ik.cpp:
```diff
@@ -362,6 +373,230 @@ bool SNS_IK::nullspaceBiasTask(const KDL::JntArray& q_bias,
+    }
+    task2.desired = Eigen::VectorXd::Zero(q_bias.rows());
+    for (size_t ii = 0; ii < q_bias.rows(); ++ii) {
+      // This calculates a "nullspace velocity".
```
This was copied from the Velocity IK solver. Does this need to be re-formulated for a Acceleration IK solver? Same thing for the next secondary task.

robodreamer replied:
Yes, Good point. The name could now be changed to nullspace acceleration although the calculation stays the same since the desired acceleration is zero.

forrest-rm commented on sns_ik_lib/src/sns_ik_base.cpp:
```diff
@@ -189,9 +189,12 @@ SnsIkBase::ExitCode SnsIkBase::computeTaskScalingFactor(const Eigen::MatrixXd& J
   for (int i = 0; i < nJnt_; i++) {
     if (jntIsFree[i]) {
       jntScaleFactorArr(i) = SnsIkBase::findScaleFactor(lowMargin(i), uppMargin(i), a(i));
+      if (jntScaleFactorArr(i) == 1 && (jointOut(i) < (qLow_(i) - SnsIkBase::BOUND_TOLERANCE) || jointOut(i) > (qUpp_(i) + SnsIkBase::BOUND_TOLERANCE)))
```
Check line length

forrest-rm commented on sns_ik_lib/src/sns_ik_base.cpp:
```diff
@@ -189,9 +189,12 @@ SnsIkBase::ExitCode SnsIkBase::computeTaskScalingFactor(const Eigen::MatrixXd& J
   Eigen::ArrayXd jntScaleFactorArr(nJnt_);
```
What does "Arr" stand for?

robodreamer replied:
It just indicates that the variable type is an Eigen array instead of a vector. It is the convention that was used in Matt's initial sns ik code.

forrest-rm commented on sns_ik_lib/src/sns_ik_base.cpp:
```diff
@@ -189,9 +189,12 @@ SnsIkBase::ExitCode SnsIkBase::computeTaskScalingFactor(const Eigen::MatrixXd& J
     if (jntIsFree[i]) {
       jntScaleFactorArr(i) = SnsIkBase::findScaleFactor(lowMargin(i), uppMargin(i), a(i));
+      if (jntScaleFactorArr(i) == 1 && (jointOut(i) < (qLow_(i) - SnsIkBase::BOUND_TOLERANCE) || jointOut(i) > (qUpp_(i) + SnsIkBase::BOUND_TOLERANCE)))
+        jntScaleFactorArr(i) = 1e-3;
```
Magic number

forrest-rm commented on sns_ik_lib/test/sns_ik_acc_test.cpp:
```diff
@@ -0,0 +1,327 @@
+      results.solveTime += tmp.solveTime;
+      results.fkLinErr += tmp.fkLinErr;
+      results.fkAngErr += tmp.fkAngErr;
+      ASSERT_LE(tmp.fkLinErr, ACC_IK_TEST_FK_LIN_TOL);
```
Should these asserts also happen when iTest == 0? If so, move them out of the else statement.

forrest-rm commented on sns_ik_lib/test/sns_ik_acc_test.cpp:
```diff
@@ -0,0 +1,327 @@
+  // Run the test:
+  /*
+   * Note: This seems to cause a segfault in ubuntu 16.04 with Eigen 3.3.4, but it
+   * works fine in Ubuntu 14.04 with Eigen 3.2.0. FIXME
```
Is there a issue on github for this?

robodreamer replied:
I just created an issue for this.

forrest-rm commented on sns_ik_lib/test/sns_ik_acc_test.cpp:
```diff
@@ -0,0 +1,327 @@
+ *************************************************************************************************/
+
+/*
+ * Run the benchmark test on all five versions of the acceleration solver.
```
There's only one acc solver right now....

robodreamer replied:
Yep. Right..

forrest-rm reviewed (commented):
Overall, it looks mostly good.

Commit:
Addressed the reviewers' comments

- removed unnecessary variables in jacobian dot computation
- removed hard-coded numbers
- enabled sns acc ik by default (only called upon request)
- other cleanups

forrest-rm commented on sns_ik_lib/src/sns_acceleration_ik.cpp:
```diff
@@ -126,10 +128,18 @@ void SNSAccelerationIK::shapeJointAccelerationBound(const Eigen::VectorXd &actua
+    // check whether the acceleration limits are correctly shaped
+    if (ddotQmin(i) > ddotQmax(i)) {
+      if (ddotQmin(i) > 0) {
+      ROS_ERROR_STREAM("Lower limit has been reversed for J"<< i <<"!\n"<< "jointLimit_Low: " << jointLimit_low(i) 
```
Need additional indentation

forrest-rm replied:
Should this return an error code and/or revert to the previous values?

forrest-rm commented on sns_ik_lib/src/sns_ik.cpp:
```diff
@@ -28,6 +28,10 @@
 
+  // define some constants
+  static const size_t CARTESIAN_DOF = 6;
+  static const size_t NUM_LINKS = 9; // the number of links in robot kinematic chain
```
Why is this hard-coded and will this break things if the the DOF is different?

forrest-rm reviewed (commented):
Overall, it looks fairly good. I just had a few questions.
